src/batch-transaction.js

```diff
@@ -0,0 +1,343 @@
+ * @typedef {array} CreateQueryPartitionsResponse
+ * @property {QueryPartition[]} 0 List of query partitions.
+ * @property {object} 1 The full API response.
+ */
+/**
+ * @callback CreateQueryPartitionsCallback
+ * @param {?Error} err Request error, if any.
+ * @param {QueryPartition[]} partitions List of query partitions.
+ * @param {object} apiResponse The full API response.
+ */
+/**
+ * Creates a set of query partitions that can be used to execute a query
+ * operation in parallel. Partitions become invalid when the transaction used
+ * to create them is closed.
+ *
+ * @param {string|object} query A SQL query or
+ *     [`ExecuteSqlRequest`](https://cloud.google.com/spanner/docs/reference/rpc/google.spanner.v1#google.spanner.v1.ExecuteSqlRequest)
+ *     object.
+ * @param {object} [query.params] A map of parameter name to values.
+ * @param {object} [query.partitionOptions] A map of partition options.
+ * @param {object} [query.types] A map of parameter types.
+ * @param {CreateQueryPartitionsCallback} [callback] Callback function.
+ * @returns {Promise<CreateQueryPartitionsResponse>}
+ *
+ * @example <caption>include:samples/batch.js</caption>
+ * region_tag:create_query_partitions
+ */
+BatchTransaction.prototype.createQueryPartitions = function(query, callback) {
+  if (is.string(query)) {
+    query = {
+      sql: query,
+    };
+  }
+
+  var reqOpts = codec.encodeQuery(query);
+  var gaxOpts = query.gaxOptions;
+
+  if (gaxOpts) {
+    delete reqOpts.gaxOptions;
+  }
+
+  this.createPartitions_(
+    {
+      client: 'SpannerClient',
+      method: 'partitionQuery',
+      reqOpts,
+      gaxOpts,
+    },
+    callback
+  );
+};
+
+/**
+ * Generic create partition method. Handles common parameters used in both
+ * {@link BatchTransaction#createQueryPartitions} and {@link BatchTransaction#createReadPartitions}
+ *
+ * @private
+ *
+ * @param {object} config The request config.
+ * @param {function} Callback function.
+ */
+BatchTransaction.prototype.createPartitions_ = function(config, callback) {
+  var self = this;
+
+  var query = extend({}, config.reqOpts, {
+    session: this.session.formattedName_,
+    transaction: {id: this.id},
+  });
+
+  config.reqOpts = extend({}, query);
+  delete query.partitionOptions;
+
+  this.request(config, function(err, resp) {
+    if (err) {
+      callback(err, null, resp);
+      return;
+    }
+
+    var partitions = resp.partitions.map(function(partition) {
+      return extend({}, query, partition);
+    });
+
+    if (resp.transaction) {
+      self.id = resp.transaction.id;
+      self.readTimestamp = resp.transaction.readTimestamp;
+    }
+
+    callback(null, partitions, resp);
+  });
+};
+
+/**
+ * @typedef {object} ReadPartition
+ * @mixes ReadRequestOptions
+ * @property {string} partitionToken The partition token.
+ */
+/**
+ * @typedef {array} CreateReadPartitionsResponse
+ * @property {ReadPartition[]} 0 List of read partitions.
+ * @property {object} 1 The full API response.
```

Comment:
No API response from GAPIC.

Reply:
See https://github.com/googleapis/nodejs-spanner/pull/132#discussion_r171338900

Reply:
https://github.com/callmehiphop/nodejs-spanner/blob/df0116e5f450cc8bf4759e49494b08b92dc83634/src/v1/spanner_client.js#L1288

```
  * @param {function(?Error, ?Object)} [callback]
   *   The function which will be called with the result of the API call.
   *
   *   The second parameter to the callback is an object representing [PartitionResponse]{@link google.spanner.v1.PartitionResponse}.
```

Reply:
So then are you asking me to rename api response to partition response?

Reply:
I might just not get JSDocs, but it looks like this says that there are two responses that come back, an array of "ReadPartition"s, and an API response. But in the code, we only return what GAPIC returns, which is just "PartitionResponse".

Reply:
Oh, the method we call here isn't a gapic method, this is a private one we control. https://github.com/googleapis/nodejs-spanner/pull/132/files/df0116e5f450cc8bf4759e49494b08b92dc83634#diff-2577c1ac4653d72633571d9a02d071d8R180

Reply:
Impressive patience you have with me.

*/mark as resolved*